lib/contentful_delivery/assets.ex

```diff
@@ -0,0 +1,188 @@
+  @moduledoc """
+  Deals with the loading of assets from a given `Contentful.Space`
+
+  https://www.contentful.com/developers/docs/references/content-delivery-api/#/reference/assets
+  """
+
+  alias Contentful.{Asset, Collection, CollectionStream, Delivery, Space}
+
+  @behaviour Collection
+  @behaviour CollectionStream
+  @doc """
+  fetches one asset from a given space
+
+  ## Examples
+      space = "my_space_id"
+      {:ok, %Asset{ meta_data: %{ id: "<asset_id>"}} = asset}
+        =  Assets.fetch_one("<asset_id>", space)
+
+      # using the configured space from config/config.exs
+      {:ok, %Asset{ meta_data: %{ id: "<asset_id>"}} = asset}
+        =  Assets.fetch_one("<asset_id>")
+
+  """
+  @impl Collection
+  @spec fetch_one(
+          String.t(),
+          Space.t() | String.t(),
+          String.t() | nil,
+          String.t() | nil
+        ) ::
+          {:ok, Asset.t()}
+          | {:error, atom(), original_message: String.t()}
+          | {:error, :rate_limit_exceeded, wait_for: integer()}
+          | {:error, :unknown}
+
+  def fetch_one(asset, space \\ nil, env \\ nil, api_key \\ nil)
+
+  def fetch_one(asset_id, %Space{meta_data: %{id: space_id}}, env, api_key) do
+    space_id
+    |> build_single_request(asset_id, env, api_key)
+    |> Delivery.send_request()
+    |> Delivery.parse_response(&build_asset/1)
+  end
+
+  def fetch_one(asset_id, space_id, env, api_key) do
+    fetch_one(asset_id, %Space{meta_data: %{id: space_id}}, env, api_key)
+  end
+
+  @doc """
+  Fetches all assets for a given `Contentful.Space`.
+
+  Will take basic collection filters into account, specifically `:limit` and `:skip` to traverse and
+  limit the collection of assets.
+
+  ## Examples
+      space = "my_space_id"
+      {:ok, [%Asset{} | _]} = Assets.fetch_all([], space)
+
+      {:ok, [
+        %Asset{ meta_data: %{ id: "foobar_0"}},
+        %Asset{ meta_data: %{ id: "foobar_1"}},
+      ], total: 3} = Assets.fetch_all([limit: 2], space)
+
+
+      # using the configured space
+      {:ok, [
+        %Asset{ meta_data: %{ id: "foobar_1"}},
+        %Asset{ meta_data: %{ id: "foobar_2"}}
+      ], total: 3} = Assets.fetch_all(skip: 1)
+
+      {:ok, [
+        %Asset{ meta_data: %{ id: "foobar_0"}}
+      ], total: 3} = Assets.fetch_all(limit: 1)
+
+      {:ok, [
+        %Asset{ meta_data: %{ id: "foobar_2"}}
+      ], total: 3} = Assets.fetch_all(limit: 1, skip: 2)
+  """
+  @impl Collection
+  @spec fetch_all(
+          list(keyword()),
+          Space.t() | String.t(),
+          String.t() | nil,
+          String.t() | nil
+        ) ::
+          {:ok, list(Asset.t())}
+          | {:error, atom(), original_message: String.t()}
+          | {:error, :rate_limit_exceeded, wait_for: integer()}
+          | {:error, :unknown}
+
+  def fetch_all(
+        options \\ [],
+        space \\ Delivery.config(:space),
+        env \\ Delivery.config(:environment),
+        api_key \\ Delivery.config(:access_token)
```

Comment:
I wonder if we should pass in only the `Delivery.config()`.
Ideally the config would be its own struct module and at least provide an interface to explicitly set/override the environment (even that should probably be a future feature). I believe that this would make more sense for a start, since the common case is to use a singular space + api_key pair, and even mostly work on a single environment. So this triplet appears all the time.
The current fixed order is potentially strange for some, who want to set a different api_key, but now have to set also the environment first. Everyone will have their argument order preference, so a map/struct should avoid this problem here.
WDYT?

Reply:
The mid term plan is to migrate the Configuration into another module. Overall, the default call to most function will fall back to the configured values:

```elixir
# will fetch the space configured for the current config/config.exs
{:ok, space} = Contentful.Delivery.fetch_one()
```
I will move out the config once `Contentful.Management` is ready (when I actually need a separate config loader.

Reply:
I do plan Keyword lists at this point, merging the param lists with the `:limit`, `:order`, etc. to make the order arbitrary:

```elixir
{:ok, assets} = Contentful.Delivery.Assets.fetch_all(limit: 100, space_id: 'space_id`, environment: 'env')
```

etc.